### ISSUE TITLE
FixedSizeData usage

### ISSUE THREAD

**Issue by vfdev-5** [ppwwyyxx/tensorpack#401]
Hi,

I wonder what is a proper usage of `FixedSizeData` ? 
For example, taking Cifar10 dataset I want to iterate twice over a reduced Cifar10 dataset and to have both times the same images. 
```
train_ds = dataset.Cifar10('train')
train_ds = FixedSizeData(train_ds, size=12)
```
However, when I rerun the code to generate images, it returns next images. Is it a correct behavior ?
Full code is [here](https://gist.github.com/vfdev-5/0817f9fdd029ba00d6d0ad5c54597c9d)

Thank you

**Comment by ppwwyyxx** [comment]
The [document](http://tensorpack.readthedocs.io/en/latest/modules/dataflow.html#tensorpack.dataflow.FixedSizeData) says, "The iterator state of the underlying DataFlow will be kept if not exhausted". So this is actually an intended behavior.

**Comment by ppwwyyxx** [comment]
Maybe an option can be added to get a different behavior.

**Comment by vfdev-5** [comment]
Okay, I can propose a PR

**Comment by ppwwyyxx** [comment]
That's OK. But note that although you can get a subset of the dataset, you still lose the opportunity to fully shuffle the data. So maybe this is not that useful.
As it is a streaming interface, a full shuffle can only happen at the very beginning of the dataflow.

**Comment by vfdev-5** [comment]
Yes, I see what you mean. So every epoch I will get the same sequence of images. Sure that this is not good.

**Comment by vfdev-5** [comment]
@ppwwyyxx actually my initial problem before facing `FixedSizeData` is related to `PrefetchData` and what you mention in documentation : 
> e.g. you are likely to see duplicated datapoints at the beginning

It is very strange to see several identical batchs. Do you think we can do something with that or we can keep it like this ?

**Comment by ppwwyyxx** [comment]
This is the current implementation of `PrefetchData` -- it forks the dataflow. There is no way to __automatically__ make a single iterator run in parallel, unless you fork it to become many iterators.

`ThreadedMapData` also runs in parallel but doesn't do the fork. (Similar things can be done for multiprocess as well but that hasn't been implemented). But they essentially have different semantics. By forking a dataflow it accelerates that dataflow. If only running the mapping function in parallel, the underlying dataflow doesn't get accelerated.

**Comment by vfdev-5** [comment]
Hi @ppwwyyxx 

Sorry to revive this closed issue, I would like to comeback to `PrefetchData` function and duplicated datapoints. As you said, duplicates are explained by forking the dataflow, i.e. dataflow iterator.
Comparing to other frameworks, for example, *PyTorch*, their [`Dataset`](http://pytorch.org/docs/master/data.html#torch.utils.data.Dataset) implementation requires similar methods: `__len__` and `__getitem__`. Another class [`DataLoader`](http://pytorch.org/docs/master/data.html#torch.utils.data.DataLoader) provides similarly a [prefetch with multiprocessing](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py). Their implementation does not have duplicates in the beginning.
I wonder why you chose to provide iterator with `get_data` and not a similar to *pytorch* API ?

Thanks

**Comment by ppwwyyxx** [comment]
The interface is different in that we don't have `__getitem__`. DataFlow is an iterator interface, while pytorch dataset is an array interface.
An array interface is less performant because it needs to keep track of the indices of every data point.

DataFlow can emulate the array interface, by having a dataflow producing the indices, and a `MapData` to gather the actual data from the indices. This way, parallel dataflow can avoid producing duplicated data as long as they take indices from the same queue, and this is exactly how pytorch multiprocess prefetch is implemented.

To do this, you can:
```python
df = DataFlowProducingIndices()
df = PrefetchData(df, 100, 1)  # use 1 multiprocessing.Queue for indices
df = MapData(df, from_indice_to_data)
df = PrefetchDataZMQ(df, 10)   # fork the mapping part
```

**Comment by vfdev-5** [comment]
@ppwwyyxx thank you !
Actually, following your comment and the documentation [here](http://tensorpack.readthedocs.io/en/latest/tutorial/efficient-dataflow.html#sequential-read) :
```
ds = LMDBData(db, shuffle=False)
ds = LocallyShuffleData(ds, 50000)

ds = PrefetchData(ds, 5000, 1) # !!! ADDING THIS !!!

ds = LMDBDataPoint(ds)
ds = MapDataComponent(ds, lambda x: cv2.imdecode(x, cv2.IMREAD_COLOR), 0)
ds = AugmentImageComponent(ds, lots_of_augmentors)
ds = PrefetchDataZMQ(ds, 25)
ds = BatchData(ds, 256)
```
it helps also to avoid duplicates in my case too without using indices ! 

In my case it looks like :
```
train_ds = Cifar10('train')
pref_train_ds = PrefetchData(train_ds, nr_prefetch=5*batch_size, nr_proc=1)
aug_train_ds = AugmentImageComponent(pref_train_ds, geom_augmenters, index=0, copy=False)
aug_train_ds = AugmentImageComponent(aug_train_ds, color_augmenters, index=0, copy=False)  
train_batches = BatchData(aug_train_ds, batch_size=batch_size, use_list=True, remainder=False)
train_batches = PrefetchDataZMQ(train_batches, nr_proc=15)
```
Why in your examples you have (main process makes batches) :
```
... -> PrefetchDataZMQ -> BatchData
```
and not 
```
... -> BatchData -> PrefetchDataZMQ
```
is there a particular reason ?

**Comment by ppwwyyxx** [comment]
It's up to you to batch before or after prefetch.
